velox/functions/sparksql/aggregates/FirstLastAggregate.cpp

```diff
@@ -0,0 +1,353 @@
+        if (exec::Aggregate::isNull(group)) {
+          (*result)->setNull(i, true);
+        } else {
+          exec::Aggregate::clearNull(rawNulls, i);
+          auto accumulator =
+              exec::Aggregate::value<SingleValueAccumulator>(group);
+          accumulator->read(*result, i);
+        }
+      }
+    }
+  }
+
+  void extractAccumulators(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    extractValues(groups, numGroups, result);
+  }
+
+ protected:
+  void updateValue(vector_size_t i, char* group, DecodedVector& decoded) {
+    exec::Aggregate::clearNull(group);
+
+    if (numeric) {
+      auto value = decoded.valueAt<TDataType>(i);
+      *exec::Aggregate::value<TDataType>(group) = value;
+    } else {
+      const auto* indices = decoded.indices();
+      const auto* baseVector = decoded.base();
+      auto* accumulator = exec::Aggregate::value<SingleValueAccumulator>(group);
+      accumulator->write(baseVector, indices[i], exec::Aggregate::allocator_);
+    }
+  }
+};
+
+template <bool numeric, bool ignoreNull, typename T>
+class FirstAggregate : public FirstLastAggregateBase<numeric, T> {
+ public:
+  explicit FirstAggregate(TypePtr resultType)
+      : FirstLastAggregateBase<numeric, T>(resultType) {}
+
+  void addRawInput(
+      char** groups,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /* mayPushdown */) override {
+    DecodedVector decoded(*args[0], rows);
+
+    std::vector<vector_size_t> selectedIndexes;
+    selectedIndexes.reserve(rows.countSelected());
+    rows.applyToSelected(
+        [&selectedIndexes](vector_size_t i) { selectedIndexes.push_back(i); });
+
+    if (decoded.isConstantMapping()) {
+      if (!decoded.isNullAt(0)) {
+        rows.applyToSelected(
+            [&](vector_size_t i) { this->updateValue(i, groups[i], decoded); });
+      }
+    } else if (decoded.mayHaveNulls()) {
+      for (auto it = selectedIndexes.crbegin(); it < selectedIndexes.crend();
+           it++) {
+        if (!decoded.isNullAt(*it)) {
+          this->updateValue(*it, groups[*it], decoded);
+        } else if (!ignoreNull) {
+          exec::Aggregate::setNull(groups[*it]);
+        }
+      }
+    } else {
+      for (auto it = selectedIndexes.crbegin(); it < selectedIndexes.crend();
+           it++) {
+        this->updateValue(*it, groups[*it], decoded);
+      }
+    }
+  }
+
+  void addIntermediateResults(
+      char** groups,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool mayPushdown) override {
+    addRawInput(groups, rows, args, mayPushdown);
+  }
+
+  void addSingleGroupRawInput(
+      char* group,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /* mayPushdown */) override {
+    DecodedVector decoded(*args[0], rows);
+
+    auto i = rows.begin();
+    while (i < rows.end()) {
+      if (!rows.isValid(i)) {
```

Comment:
use rows.applyToSelected

Reply:
BTW, this logic is comes from old `LastAggregate.cpp`, I think it's better than `rows.applyToSelected`

Reply:
BTW, if you want to implement early return, then use rows.testSelected.

Reply:
> What happens if we use rows.applyToSelected here?

It's fine, I just want early return like you said, I will try to use `rows.testSelected`, thanks!